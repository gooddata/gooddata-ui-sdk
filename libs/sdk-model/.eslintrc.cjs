// (C) 2020 GoodData Corporation
module.exports = {
    parser: "@typescript-eslint/parser",
    plugins: ["prettier", "sonarjs", "eslint-plugin-tsdoc", "regexp"],
    extends: [
        "@gooddata",
        "plugin:import-esm/recommended",
        "plugin:sonarjs/recommended",
        "plugin:regexp/recommended",
        "../../.eslintrc.js",
    ],
    parserOptions: { tsconfigRootDir: __dirname },
};
