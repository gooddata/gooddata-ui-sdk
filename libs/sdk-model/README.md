# GoodData.UI domain models

[![npm version](https://img.shields.io/npm/v/@gooddata/sdk-model)](https://www.npmjs.com/@gooddata/sdk-model)&nbsp;
[![npm monthly downloads](https://img.shields.io/npm/dm/@gooddata/sdk-model)](https://npmcharts.com/compare/@gooddata/sdk-model?minimal=true)&nbsp;
![typescript](https://img.shields.io/badge/typescript-first-blue?logo=typescript)

This package is a part of the [GoodData.UI SDK](https://sdk.gooddata.com/gooddata-ui/docs/about_gooddataui.html).
To learn more, check [the source monorepo](https://github.com/gooddata/gooddata-ui-sdk).

This package provides domain models for GoodData.UI.

These domain models are backend-agnostic. This makes them reusable across different [Analytical Backend](https://www.npmjs.com/package/@gooddata/sdk-backend-spi) implementations.
The package includes TypeScript type definitions, factory functions, functions to get or set certain properties
of the objects in an immutable way, and more. These are used in both the `@gooddata/sdk-backend-*` and `@gooddata/sdk-ui* packages`.

## License

(C) 2017-2022 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE).
