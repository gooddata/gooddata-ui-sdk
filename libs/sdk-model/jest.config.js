// (C) 2019 GoodData Corporation
import base from "../../common/config/jest/jest.config.base.mjs";

export default {
  ...base,
};