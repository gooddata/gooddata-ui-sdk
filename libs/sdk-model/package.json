{
    "name": "@gooddata/sdk-model",
    "version": "8.12.3-beta.0",
    "author": "GoodData",
    "description": "GoodData Model definitions used by UI components and Backend SPI and its implementations",
    "repository": {
        "type": "git",
        "url": "https://github.com/gooddata/gooddata-ui-sdk.git",
        "directory": "libs/sdk-model"
    },
    "license": "MIT",
    "main": "dist/index.js",
    "module": "esm/index.js",
    "es2015": "esm/index.js",
    "browser": "dist/index.js",
    "typings": "esm/index.d.ts",
    "sideEffects": false,
    "files": [
        "dist/**/*.js",
        "dist/**/*.json",
        "dist/**/*.d.ts",
        "dist/**/*.map",
        "esm/**/*.js",
        "esm/**/*.json",
        "esm/**/*.d.ts",
        "esm/**/*.map"
    ],
    "config": {
        "eslint": "-c .eslintrc.js --ext ts src/"
    },
    "scripts": {
        "clean": "rm -rf ci dist esm coverage *.log && jest --clearCache",
        "dev": "tsc -p tsconfig.dev.json --watch",
        "build": "concurrently \"npm run build-cjs\" \"npm run build-esm\" && npm run api-extractor",
        "build-cjs": "tsc -p tsconfig.build.json",
        "build-esm": "tsc -p tsconfig.build.esm.json",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest --maxWorkers=${JEST_MAX_WORKERS:-'45%'}",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --ci --config jest.ci.js",
        "eslint": "eslint $npm_package_config_eslint",
        "eslint-ci": "mkdir -p ./ci/results && eslint -f checkstyle -o ci/results/eslint-results.xml $npm_package_config_eslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "dep-cruiser": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "dep-cruiser-ci": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "validate": "npm run dep-cruiser && npm run eslint && npm run prettier-check",
        "validate-ci": "npm run dep-cruiser-ci && npm run eslint-ci && npm run prettier-check"
    },
    "dependencies": {
        "json-stable-stringify": "^1.0.1",
        "lodash": "^4.17.19",
        "spark-md5": "^3.0.0",
        "stringify-object": "^3.3.0",
        "ts-invariant": "^0.7.3",
        "tslib": "^2.0.0"
    },
    "devDependencies": {
        "@gooddata/eslint-config": "^4.1.0",
        "@microsoft/api-documenter": "^7.17.0",
        "@microsoft/api-extractor": "^7.20.0",
        "@types/jest": "^27.0.1",
        "@types/json-stable-stringify": "^1.0.32",
        "@types/lodash": "^4.14.158",
        "@types/spark-md5": "^3.0.1",
        "@types/stringify-object": "^3.2.0",
        "@typescript-eslint/eslint-plugin": "^5.45.0",
        "@typescript-eslint/parser": "^5.45.0",
        "concurrently": "^6.0.2",
        "dependency-cruiser": "^12.12.2",
        "eslint": "^8.25.0",
        "eslint-plugin-header": "^3.0.0",
        "eslint-plugin-import": "^2.22.0",
        "eslint-plugin-jest": "^27.1.1",
        "eslint-plugin-no-only-tests": "^2.4.0",
        "eslint-plugin-prettier": "^4.0.0",
        "eslint-plugin-regexp": "^1.9.0",
        "eslint-plugin-sonarjs": "^0.16.0",
        "eslint-plugin-tsdoc": "^0.2.14",
        "jest": "^27.5.1",
        "jest-junit": "^3.0.0",
        "prettier": "~2.5.0",
        "ts-jest": "^27.0.5",
        "typescript": "4.0.2"
    }
}
