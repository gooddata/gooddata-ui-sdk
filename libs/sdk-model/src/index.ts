// (C) 2019-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
export { DateAttributeGranularity, DateGranularity, AllTimeGranularity } from "./base/dateGranularities";

export {
    IAttribute,
    isAttribute,
    attributeLocalId,
    AttributePredicate,
    anyAttribute,
    idMatchAttribute,
    attributesFind,
    attributeUri,
    attributeIdentifier,
    attributeAlias,
    attributeDisplayFormRef,
} from "./execution/attribute";

export {
    newAttribute,
    modifyAttribute,
    AttributeBuilder,
    AttributeModifications,
    AttributeBuilderInput,
} from "./execution/attribute/factory";

export {
    ObjectType,
    Identifier,
    Uri,
    UriRef,
    IdentifierRef,
    LocalIdRef,
    ObjRef,
    ObjRefInScope,
    isUriRef,
    isIdentifierRef,
    objRefToString,
    isLocalIdRef,
    areObjRefsEqual,
    isObjRef,
} from "./objRef";

export {
    IDimension,
    isDimension,
    dimensionTotals,
    DimensionItem,
    newTwoDimensional,
    newDimension,
    MeasureGroupIdentifier,
    dimensionSetTotals,
    dimensionsFindItem,
    ItemInDimension,
} from "./execution/base/dimension";

export { idRef, uriRef, localIdRef } from "./objRef/factory";

export { TotalType, ITotal, isTotal, newTotal, totalIsNative } from "./execution/base/totals";

export {
    SortDirection,
    IAttributeSortItem,
    ISortItem,
    IMeasureSortItem,
    ILocatorItem,
    IAttributeLocatorItem,
    IMeasureLocatorItem,
    isMeasureLocator,
    isAttributeLocator,
    isMeasureSort,
    isAttributeSort,
    isAttributeAreaSort,
    newMeasureSort,
    newAttributeSort,
    newAttributeAreaSort,
    newAttributeLocator,
    SortEntityIds,
    sortEntityIds,
    sortDirection,
    sortMeasureLocators,
    attributeLocatorElement,
    attributeLocatorIdentifier,
    measureLocatorIdentifier,
} from "./execution/base/sort";

export {
    IAttributeElementsByRef,
    IAttributeElementsByValue,
    IAttributeElements,
    IPositiveAttributeFilter,
    INegativeAttributeFilter,
    IAbsoluteDateFilter,
    IRelativeDateFilter,
    IMeasureValueFilter,
    IRankingFilter,
    RankingFilterOperator,
    isRankingFilter,
    IFilter,
    INullableFilter,
    IMeasureFilter,
    IDateFilter,
    IAttributeFilter,
    isAbsoluteDateFilter,
    isRelativeDateFilter,
    isAllTimeDateFilter,
    attributeElementsIsEmpty,
    attributeElementsCount,
    isPositiveAttributeFilter,
    isNegativeAttributeFilter,
    isDateFilter,
    isMeasureValueFilter,
    ComparisonConditionOperator,
    IComparisonCondition,
    IRangeCondition,
    MeasureValueFilterCondition,
    RangeConditionOperator,
    isAttributeFilter,
    isAttributeElementsByRef,
    isAttributeElementsByValue,
    isComparisonCondition,
    isComparisonConditionOperator,
    isFilter,
    isRangeCondition,
    isRangeConditionOperator,
    filterIsEmpty,
    filterAttributeElements,
    filterMeasureRef,
    filterObjRef,
    IAbsoluteDateFilterValues,
    IRelativeDateFilterValues,
    absoluteDateFilterValues,
    relativeDateFilterValues,
    measureValueFilterCondition,
    measureValueFilterMeasure,
    measureValueFilterOperator,
} from "./execution/filter";

export {
    newAbsoluteDateFilter,
    newNegativeAttributeFilter,
    newPositiveAttributeFilter,
    newRelativeDateFilter,
    newAllTimeFilter,
    newMeasureValueFilter,
    newRankingFilter,
} from "./execution/filter/factory";

export { mergeFilters } from "./execution/filter/filterMerge";

export {
    IMeasureTitle,
    IMeasureDefinitionType,
    IMeasureDefinition,
    ArithmeticMeasureOperator,
    IArithmeticMeasureDefinition,
    IPoPMeasureDefinition,
    IMeasure,
    MeasureAggregation,
    IPreviousPeriodMeasureDefinition,
    IPreviousPeriodDateDataSet,
    isMeasure,
    isSimpleMeasure,
    isAdhocMeasure,
    isPoPMeasure,
    isPreviousPeriodMeasure,
    isArithmeticMeasure,
    isMeasureDefinition,
    isPoPMeasureDefinition,
    isPreviousPeriodMeasureDefinition,
    isArithmeticMeasureDefinition,
    measureLocalId,
    MeasurePredicate,
    anyMeasure,
    idMatchMeasure,
    measureDoesComputeRatio,
    measureItem,
    measureUri,
    measureIdentifier,
    measureMasterIdentifier,
    measureArithmeticOperands,
    measureAlias,
    measureTitle,
    measureArithmeticOperator,
    measureFormat,
    isMeasureFormatInPercent,
    measureAggregation,
    measureFilters,
    measurePopAttribute,
    measurePreviousPeriodDateDataSets,
    MeasureOrLocalId,
} from "./execution/measure";

export {
    IPreviousPeriodDateDataSetSimple,
    ArithmeticMeasureBuilder,
    MeasureBuilder,
    MeasureModifications,
    PoPMeasureBuilder,
    PreviousPeriodMeasureBuilder,
    MeasureBuilderBase,
    newMeasure,
    modifyMeasure,
    modifySimpleMeasure,
    modifyPopMeasure,
    modifyPreviousPeriodMeasure,
    newArithmeticMeasure,
    newPopMeasure,
    newPreviousPeriodMeasure,
    MeasureEnvelope,
    ArithmeticMeasureBuilderInput,
    PoPMeasureBuilderInput,
    PreviousPeriodMeasureBuilderInput,
} from "./execution/measure/factory";

export {
    IAttributeOrMeasure,
    IBucket,
    isBucket,
    idMatchBucket,
    anyBucket,
    MeasureInBucket,
    AttributeInBucket,
    newBucket,
    bucketIsEmpty,
    bucketAttributes,
    bucketAttribute,
    bucketAttributeIndex,
    bucketMeasure,
    bucketMeasureIndex,
    bucketMeasures,
    bucketTotals,
    bucketSetTotals,
    bucketItems,
    BucketPredicate,
    applyRatioRule,
    ComputeRatioRule,
    disableComputeRatio,
    BucketItemModifications,
    BucketItemReducer,
    bucketModifyItems,
    bucketItemReduce,
} from "./execution/buckets";

export {
    bucketsFind,
    bucketsMeasures,
    bucketsIsEmpty,
    bucketsAttributes,
    bucketsFindMeasure,
    bucketsById,
    bucketsFindAttribute,
    bucketsItems,
    bucketsTotals,
    bucketsModifyItem,
    bucketsReduceItem,
} from "./execution/buckets/bucketArray";

export { bucketItemLocalId } from "./execution/buckets/bucketItem";

export {
    IExecutionDefinition,
    IExecutionConfig,
    DimensionGenerator,
    defWithFilters,
    defFingerprint,
    defSetDimensions,
    defSetSorts,
    defTotals,
    defSetExecConfig,
    IPostProcessing,
    defSetPostProcessing,
} from "./execution/executionDefinition";

export {
    newDefForItems,
    newDefForBuckets,
    newDefForInsight,
    defWithDimensions,
    defWithSorting,
    defWithPostProcessing,
    defWithDateFormat,
    defWithExecConfig,
    defaultDimensionsGenerator,
    emptyDef,
} from "./execution/executionDefinition/factory";

export {
    GuidType,
    RgbType,
    IRgbColorValue,
    IColor,
    IColorPalette,
    IColorPaletteItem,
    IColorFromPalette,
    IRgbColor,
    isColorFromPalette,
    isRgbColor,
    colorPaletteItemToRgb,
    colorPaletteToColors,
} from "./colors";

export {
    IInsight,
    IInsightDefinition,
    IVisualizationClass,
    VisualizationProperties,
    IColorMappingItem,
    isInsight,
    insightRef,
    insightId,
    insightItems,
    insightMeasures,
    insightHasMeasures,
    insightAttributes,
    insightHasAttributes,
    insightHasDataDefined,
    insightProperties,
    insightBuckets,
    insightSorts,
    insightBucket,
    insightTitle,
    insightUri,
    insightIsLocked,
    insightCreated,
    insightUpdated,
    insightTotals,
    insightFilters,
    insightVisualizationUrl,
    insightSetFilters,
    insightSetBuckets,
    insightSetProperties,
    insightSetSorts,
    insightModifyItems,
    insightReduceItems,
    visClassUrl,
    visClassId,
    visClassUri,
} from "./insight";

export { newInsightDefinition, InsightDefinitionBuilder, InsightModifications } from "./insight/factory";

export { insightSanitize } from "./insight/sanitization";

export { factoryNotationFor } from "./execution/objectFactoryNotation";
