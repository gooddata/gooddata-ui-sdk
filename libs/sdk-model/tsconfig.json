{
    "compilerOptions": {
        "target": "es2017",
        "lib": ["es2017", "dom"],
        "module": "NodeNext",
        "moduleResolution": "nodenext",
        "esModuleInterop": true,
        "importHelpers": true,        
        "sourceMap": true,
        "outDir": "./esm",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "removeComments": false
    },

    "exclude": ["**/node_modules","esm"]
}
