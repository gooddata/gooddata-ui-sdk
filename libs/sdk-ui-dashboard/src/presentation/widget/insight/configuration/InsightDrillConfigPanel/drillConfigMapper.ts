// (C) 2022 GoodData Corporation
import {
    IDrillToDashboard,
    IDrillToInsight,
    InsightDrillDefinition,
    isDrillFromAttribute,
    isDrillToAttributeUrl,
    isDrillToCustomUrl,
    isDrillToDashboard,
    isDrillToInsight,
    ObjRef,
    objRefToString,
} from "@gooddata/sdk-model";
import { IAvailableDrillTargets } from "@gooddata/sdk-ui";
import { defineMessage } from "react-intl";
import {
    getDrillOriginLocalIdentifier,
    getLocalIdentifierOrDie,
    getValidDrillOriginAttributes,
} from "../../../../../_staging/drills/drillingUtils.js";
import {
    DRILL_TARGET_TYPE,
    IDrillConfigItem,
    IDrillConfigItemBase,
    IDrillDownAttributeHierarchyConfig,
    IDrillToDashboardConfig,
    IDrillToInsightConfig,
    IDrillToUrl,
    IDrillToUrlConfig,
    isDrillToUrl,
    UrlDrillTarget,
} from "../../../../drill/types.js";
import { IGlobalDrillDownAttributeHierarchyDefinition } from "../../../../../types.js";

function getTitleFromDrillableItemPushData(items: IAvailableDrillTargets, itemId: string): string {
    const measureItems = items.measures || [];
    const measureResult = measureItems.find(
        (measureResult) => measureResult.measure.measureHeaderItem.localIdentifier === itemId,
    );

    if (measureResult) {
        return measureResult.measure.measureHeaderItem.name;
    }

    const attributeItems = items.attributes || [];
    const attributeResult = attributeItems.find(
        (attributeItem) => attributeItem.attribute.attributeHeader.localIdentifier === itemId,
    );

    return attributeResult ? attributeResult.attribute.attributeHeader.formOf.name : "";
}

const buildUrlDrillTarget = (drillData: IDrillToUrl): UrlDrillTarget => {
    if (isDrillToCustomUrl(drillData)) {
        const customUrl = drillData.target.url;
        return {
            customUrl,
        };
    }
    if (isDrillToAttributeUrl(drillData)) {
        const {
            displayForm: insightAttributeDisplayForm,
            hyperlinkDisplayForm: drillToAttributeDisplayForm,
        } = drillData.target;

        return {
            insightAttributeDisplayForm,
            drillToAttributeDisplayForm,
        };
    }
    throw new Error("Unsupported URL drill type!");
};

const createInsightConfig = (
    drillData: IDrillToInsight,
    supportedItemsForWidget: IAvailableDrillTargets,
    widgetRef: ObjRef,
): IDrillToInsightConfig => {
    const originLocalIdentifier = isDrillFromAttribute(drillData.origin)
        ? getLocalIdentifierOrDie(drillData.origin?.attribute)
        : getLocalIdentifierOrDie(drillData.origin?.measure);

    return {
        type: isDrillFromAttribute(drillData.origin) ? "attribute" : "measure",
        originLocalIdentifier,
        title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
        attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
        localIdentifier: drillData.localIdentifier!,
        drillTargetType: DRILL_TARGET_TYPE.DRILL_TO_INSIGHT,
        insightRef: drillData.target,
        complete: true,
        widgetRef: widgetRef,
    };
};

const createDashboardConfig = (
    drillData: IDrillToDashboard,
    supportedItemsForWidget: IAvailableDrillTargets,
    widgetRef: ObjRef,
): IDrillToDashboardConfig => {
    const originLocalIdentifier = isDrillFromAttribute(drillData.origin)
        ? getLocalIdentifierOrDie(drillData.origin?.attribute)
        : getLocalIdentifierOrDie(drillData.origin?.measure);

    return {
        type: isDrillFromAttribute(drillData.origin) ? "attribute" : "measure",
        originLocalIdentifier,
        title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
        attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
        localIdentifier: drillData.localIdentifier!,
        drillTargetType: DRILL_TARGET_TYPE.DRILL_TO_DASHBOARD,
        dashboard: drillData.target,
        complete: true,
        widgetRef: widgetRef,
    };
};

const invalidUrlMessage = defineMessage({
    id: "configurationPanel.drillConfig.drillIntoUrl.invalidCustomUrl",
});

const createUrlConfig = (
    drillData: IDrillToUrl,
    supportedItemsForWidget: IAvailableDrillTargets,
    invalidCustomUrlDrillLocalIds: string[],
    widgetRef: ObjRef,
): IDrillToUrlConfig => {
    const originLocalIdentifier = getDrillOriginLocalIdentifier(drillData);

    const hasWarning = invalidCustomUrlDrillLocalIds.includes(originLocalIdentifier);

    return {
        type: isDrillFromAttribute(drillData.origin) ? "attribute" : "measure",
        originLocalIdentifier,
        title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
        attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
        localIdentifier: drillData.localIdentifier!,
        drillTargetType: DRILL_TARGET_TYPE.DRILL_TO_URL,
        urlDrillTarget: buildUrlDrillTarget(drillData),
        complete: true,
        warning: hasWarning ? invalidUrlMessage.id : undefined,
        widgetRef: widgetRef,
    };
};

const createImplicitConfig = (): IDrillConfigItemBase => {
    return null as unknown as IDrillConfigItemBase;
};

const createConfig = (
    drillData: InsightDrillDefinition,
    supportedItemsForWidget: IAvailableDrillTargets,
    invalidCustomUrlDrillLocalIds: string[],
    widgetRef: ObjRef,
) => {
    if (isDrillToInsight(drillData)) {
        return createInsightConfig(drillData, supportedItemsForWidget, widgetRef);
    }

    if (isDrillToDashboard(drillData)) {
        return createDashboardConfig(drillData, supportedItemsForWidget, widgetRef);
    }

    if (isDrillToUrl(drillData)) {
        return createUrlConfig(drillData, supportedItemsForWidget, invalidCustomUrlDrillLocalIds, widgetRef);
    }

    return createImplicitConfig();
};

/**
 * @internal
 */
export const getMappedConfigForWidget = (
    configForWidget: InsightDrillDefinition[],
    supportedItemsForWidget: IAvailableDrillTargets,
    invalidCustomUrlDrillLocalIds: string[],
    widgetRef: ObjRef,
): IDrillConfigItem[] => {
    return configForWidget.map((item) =>
        createConfig(item, supportedItemsForWidget, invalidCustomUrlDrillLocalIds, widgetRef),
    );
};

/**
 * @internal
 */
export const getGlobalDrillDownMappedConfigForWidget = (
    widgetGlobalDrillDowns: IGlobalDrillDownAttributeHierarchyDefinition[],
    supportedItemsForWidget: IAvailableDrillTargets,
    widgetRef: ObjRef,
): IDrillDownAttributeHierarchyConfig[] => {
    return widgetGlobalDrillDowns.map((globalDrillDown) => {
        const originLocalIdentifier = getDrillOriginLocalIdentifier(globalDrillDown);
        const localIdentifier = `${originLocalIdentifier}_${objRefToString(globalDrillDown.target)}`;

        return {
            type: "attribute",
            drillTargetType: DRILL_TARGET_TYPE.DRILL_DOWN,
            attributeHierarchyRef: globalDrillDown.target,
            originLocalIdentifier: originLocalIdentifier,
            title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
            attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
            localIdentifier,
            complete: true,
            widgetRef: widgetRef,
        };
    });
};
