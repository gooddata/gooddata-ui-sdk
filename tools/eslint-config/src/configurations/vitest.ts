// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
