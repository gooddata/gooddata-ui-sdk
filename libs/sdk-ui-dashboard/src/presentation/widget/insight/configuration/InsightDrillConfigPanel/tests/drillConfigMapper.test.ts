// (C) 2023 GoodData Corporation
import { describe, it, expect } from "vitest";

import { getGlobalDrillDownMappedConfigForWidget } from "../drillConfigMapper.js";
import { availableDrillTargets, globalDrillDowns } from "./drillConfigMapper.fixture.js";
import { idRef } from "@gooddata/sdk-model";

describe("drillConfigMapper", () => {
    describe("getGlobalDrillDownMappedConfigForWidget", () => {
        it("should map global drill downs to drill down attribute hierarchy configs", () => {
            expect(
                getGlobalDrillDownMappedConfigForWidget(
                    globalDrillDowns,
                    availableDrillTargets,
                    idRef("test_widget_id", "insight"),
                ),
            ).toMatchSnapshot();
        });
    });
});
