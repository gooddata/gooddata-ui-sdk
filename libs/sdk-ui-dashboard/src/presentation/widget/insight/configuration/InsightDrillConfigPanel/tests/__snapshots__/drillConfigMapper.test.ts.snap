// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`drillConfigMapper > getGlobalDrillDownMappedConfigForWidget > should map global drill downs to drill down attribute hierarchy configs 1`] = `
[
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
      "type": "attributeHierarchy",
    },
    "attributes": [
      {
        "attributeHeader": {
          "formOf": {
            "identifier": "f_owner.region_id",
            "name": "Region",
            "ref": {
              "identifier": "f_owner.region_id",
              "type": "attribute",
            },
            "uri": "",
          },
          "identifier": "f_owner.region_id",
          "localIdentifier": "458cb8556b604396be5d736ada2aa833",
          "name": "Region",
          "ref": {
            "identifier": "f_owner.region_id",
            "type": "displayForm",
          },
          "totalItems": [],
          "uri": "",
        },
      },
    ],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "458cb8556b604396be5d736ada2aa833_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
    "originLocalIdentifier": "458cb8556b604396be5d736ada2aa833",
    "title": "Region",
    "type": "attribute",
    "widgetRef": {
      "identifier": "test_widget_id",
      "type": "insight",
    },
  },
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
      "type": "attributeHierarchy",
    },
    "attributes": [
      {
        "attributeHeader": {
          "formOf": {
            "identifier": "f_owner.region_id",
            "name": "Region",
            "ref": {
              "identifier": "f_owner.region_id",
              "type": "attribute",
            },
            "uri": "",
          },
          "identifier": "f_owner.region_id",
          "localIdentifier": "458cb8556b604396be5d736ada2aa833",
          "name": "Region",
          "ref": {
            "identifier": "f_owner.region_id",
            "type": "displayForm",
          },
          "totalItems": [],
          "uri": "",
        },
      },
      {
        "attributeHeader": {
          "formOf": {
            "identifier": "f_owner.department_id",
            "name": "Department",
            "ref": {
              "identifier": "f_owner.department_id",
              "type": "attribute",
            },
            "uri": "",
          },
          "identifier": "f_owner.department_id",
          "localIdentifier": "f20e093c6af5439dbfce10ee37620fc8",
          "name": "Department",
          "ref": {
            "identifier": "f_owner.department_id",
            "type": "displayForm",
          },
          "totalItems": [],
          "uri": "",
        },
      },
    ],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "f20e093c6af5439dbfce10ee37620fc8_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
    "originLocalIdentifier": "f20e093c6af5439dbfce10ee37620fc8",
    "title": "Department",
    "type": "attribute",
    "widgetRef": {
      "identifier": "test_widget_id",
      "type": "insight",
    },
  },
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
      "type": "attributeHierarchy",
    },
    "attributes": [
      {
        "attributeHeader": {
          "formOf": {
            "identifier": "f_owner.region_id",
            "name": "Region",
            "ref": {
              "identifier": "f_owner.region_id",
              "type": "attribute",
            },
            "uri": "",
          },
          "identifier": "f_owner.region_id",
          "localIdentifier": "458cb8556b604396be5d736ada2aa833",
          "name": "Region",
          "ref": {
            "identifier": "f_owner.region_id",
            "type": "displayForm",
          },
          "totalItems": [],
          "uri": "",
        },
      },
    ],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "458cb8556b604396be5d736ada2aa833_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
    "originLocalIdentifier": "458cb8556b604396be5d736ada2aa833",
    "title": "Region",
    "type": "attribute",
    "widgetRef": {
      "identifier": "test_widget_id",
      "type": "insight",
    },
  },
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
      "type": "attributeHierarchy",
    },
    "attributes": [
      {
        "attributeHeader": {
          "formOf": {
            "identifier": "f_owner.region_id",
            "name": "Region",
            "ref": {
              "identifier": "f_owner.region_id",
              "type": "attribute",
            },
            "uri": "",
          },
          "identifier": "f_owner.region_id",
          "localIdentifier": "458cb8556b604396be5d736ada2aa833",
          "name": "Region",
          "ref": {
            "identifier": "f_owner.region_id",
            "type": "displayForm",
          },
          "totalItems": [],
          "uri": "",
        },
      },
      {
        "attributeHeader": {
          "formOf": {
            "identifier": "f_owner.department_id",
            "name": "Department",
            "ref": {
              "identifier": "f_owner.department_id",
              "type": "attribute",
            },
            "uri": "",
          },
          "identifier": "f_owner.department_id",
          "localIdentifier": "f20e093c6af5439dbfce10ee37620fc8",
          "name": "Department",
          "ref": {
            "identifier": "f_owner.department_id",
            "type": "displayForm",
          },
          "totalItems": [],
          "uri": "",
        },
      },
    ],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "f20e093c6af5439dbfce10ee37620fc8_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
    "originLocalIdentifier": "f20e093c6af5439dbfce10ee37620fc8",
    "title": "Department",
    "type": "attribute",
    "widgetRef": {
      "identifier": "test_widget_id",
      "type": "insight",
    },
  },
]
`;
